<html>
<head>
<title>5 Best Practices To Handle Exceptions In Java</title>
<link href="../Styles/Style.css" type="text/css" rel="stylesheet"/>
</head>
<body>
<h1 class="translated">在 Java 中处理异常的 5 个最佳实践</h1>
<blockquote>原文：<a href="https://blog.devgenius.io/5-best-practices-to-handle-exceptions-in-java-5e1534f83772?source=collection_archive---------4-----------------------#2022-07-29">https://blog.devgenius.io/5-best-practices-to-handle-exceptions-in-java-5e1534f83772?source=collection_archive---------4-----------------------#2022-07-29</a></blockquote><div><div class="fc ie if ig ih ii"/><div class="ij ik il im in"><h2 id="7c53" class="io ip iq bd b dl ir is it iu iv iw dk ix translated" aria-label="kicker paragraph">软件工程之旅- JAVA</h2><div class=""/><div class=""><h2 id="e002" class="pw-subtitle-paragraph jw iz iq bd b jx jy jz ka kb kc kd ke kf kg kh ki kj kk kl km kn dk translated">在 Java 中处理异常的 5 个最佳实践</h2></div><figure class="kp kq kr ks gt kt gh gi paragraph-image"><div role="button" tabindex="0" class="ku kv di kw bf kx"><div class="gh gi ko"><img src="../Images/9edafbb5e300ca92b228f79d08a1b42a.png" data-original-src="https://miro.medium.com/v2/resize:fit:1400/0*b9YBhsujeQP2mWlU"/></div></div><figcaption class="la lb gj gh gi lc ld bd b be z dk translated"><a class="ae le" href="https://unsplash.com/@ayeltvanveen" rel="noopener ugc nofollow" target="_blank">照片由 Ayelt van Veen 在 Unsplash 上拍摄</a></figcaption></figure><h1 id="e0ce" class="lf lg iq bd lh li lj lk ll lm ln lo lp kf lq kg lr ki ls kj lt kl lu km lv lw bi translated">概观</h1><p id="6fd3" class="pw-post-body-paragraph lx ly iq lz b ma mb ka mc md me kd mf mg mh mi mj mk ml mm mn mo mp mq mr ms ij bi translated">异常是在应用程序执行过程中发生的不想要的或意外的事件。异常，由程序员或软件工程师处理。Java 中的异常处理是处理运行时错误的强大机制之一，因此可以维护应用程序的正常流程，而不会出现任何问题。异常发生的原因有很多，但我可以尝试列出一些常见的运行时错误:</p><ul class=""><li id="a017" class="mt mu iq lz b ma mv md mw mg mx mk my mo mz ms na nb nc nd bi translated">无效的用户输入</li><li id="565e" class="mt mu iq lz b ma ne md nf mg ng mk nh mo ni ms na nb nc nd bi translated">由于缺少边缘情况而导致的代码错误</li><li id="9fb2" class="mt mu iq lz b ma ne md nf mg ng mk nh mo ni ms na nb nc nd bi translated">网络接线</li><li id="7f92" class="mt mu iq lz b ma ne md nf mg ng mk nh mo ni ms na nb nc nd bi translated">被遗忘</li></ul><p id="b74b" class="pw-post-body-paragraph lx ly iq lz b ma mv ka mc md mw kd mf mg nj mi mj mk nk mm mn mo nl mq mr ms ij bi translated">在本文中，我将介绍一些对您有帮助的最佳实践，并给出一些示例，这些示例可以让您更多地了解异常并将其应用到您的项目中。</p><h1 id="da32" class="lf lg iq bd lh li lj lk ll lm ln lo lp kf lq kg lr ki ls kj lt kl lu km lv lw bi translated">#1 使用 Finally 或使用 Try With Resource 语句清理资源</h1><p id="8b69" class="pw-post-body-paragraph lx ly iq lz b ma mb ka mc md me kd mf mg mh mi mj mk ml mm mn mo mp mq mr ms ij bi translated">在 Java 中，当你在你的 try 块中使用一个资源时，比如一个 InputStream，你需要在之后关闭它。在这些情况下，一个常见的错误是在 try 块的末尾关闭资源。</p><p id="8833" class="pw-post-body-paragraph lx ly iq lz b ma mv ka mc md mw kd mf mg nj mi mj mk nk mm mn mo nl mq mr ms ij bi translated">在 try 块出现异常之前，一切都会很好。那么您的资源就不能关闭，因为您的代码不能到达 try 块的末尾。为此，我们应该应用下面两种方法来使我们的资源在任何情况下都可以关闭代码。</p><h2 id="a117" class="nm lg iq bd lh nn no dn ll np nq dp lp mg nr ns lr mk nt nu lt mo nv nw lv iw bi translated">用 Finally 清理资源</h2><p id="3d76" class="pw-post-body-paragraph lx ly iq lz b ma mb ka mc md me kd mf mg mh mi mj mk ml mm mn mo mp mq mr ms ij bi translated">对于下面的例子，我们使用 finally 块来关闭我们的资源。如你所见，如果我们的代码发生了任何情况，代码总是进入 finally 块，这使得我们的资源总是像预期的那样关闭。例如，您可以参考示例代码:</p><figure class="kp kq kr ks gt kt"><div class="bz fp l di"><div class="nx ny l"/></div><figcaption class="la lb gj gh gi lc ld bd b be z dk translated">用 Finally 示例清理资源</figcaption></figure><h2 id="1f13" class="nm lg iq bd lh nn no dn ll np nq dp lp mg nr ns lr mk nt nu lt mo nv nw lv iw bi translated">通过 Try With Resource 语句清理资源</h2><p id="4a8a" class="pw-post-body-paragraph lx ly iq lz b ma mb ka mc md me kd mf mg mh mi mj mk ml mm mn mo mp mq mr ms ij bi translated">使我们的资源按预期关闭的其他方法是使用 Try With Resource 语句，如下面的代码所示。以这种方式使用时，我们的代码更少，但您的资源必须实现“自动关闭”接口。大多数 Java 标准资源都是这么做的。例如:</p><figure class="kp kq kr ks gt kt"><div class="bz fp l di"><div class="nx ny l"/></div><figcaption class="la lb gj gh gi lc ld bd b be z dk translated">通过 Try-With-Resource 语句清理资源示例</figcaption></figure><h1 id="c8c7" class="lf lg iq bd lh li lj lk ll lm ln lo lp kf lq kg lr ki ls kj lt kl lu km lv lw bi translated">#2 处理多个异常</h1><p id="6053" class="pw-post-body-paragraph lx ly iq lz b ma mb ka mc md me kd mf mg mh mi mj mk ml mm mn mo mp mq mr ms ij bi translated">在处理异常时，您可能经常面临的一个问题是如何处理 try 块的多个异常。下面这些例子可以帮助你更清楚如何得到适合你的东西。</p><h2 id="50af" class="nm lg iq bd lh nn no dn ll np nq dp lp mg nr ns lr mk nt nu lt mo nv nw lv iw bi translated">运算符“|”的多个异常</h2><p id="5719" class="pw-post-body-paragraph lx ly iq lz b ma mb ka mc md me kd mf mg mh mi mj mk ml mm mn mo mp mq mr ms ij bi translated">您可以在下面的示例中看到，我在 catch 块中使用了运算符“|”来捕捉两个异常(“ArithmeticException”和“ArrayIndexOutOfBoundsException”)。这样，万一两个异常中的任何一个发生，我都可以得到我想要的结果。</p><p id="2105" class="pw-post-body-paragraph lx ly iq lz b ma mv ka mc md mw kd mf mg nj mi mj mk nk mm mn mo nl mq mr ms ij bi translated">你也可以使用父异常来捕捉而不是使用多个异常，但是这种方式可能会让你不清楚你会发生什么异常。例如:</p><figure class="kp kq kr ks gt kt"><div class="bz fp l di"><div class="nx ny l"/></div><figcaption class="la lb gj gh gi lc ld bd b be z dk translated">运算符“|”的多个异常示例</figcaption></figure><h2 id="c8d4" class="nm lg iq bd lh nn no dn ll np nq dp lp mg nr ns lr mk nt nu lt mo nv nw lv iw bi translated">带有多个 Catch 块的多个异常</h2><p id="8402" class="pw-post-body-paragraph lx ly iq lz b ma mb ka mc md me kd mf mg mh mi mj mk ml mm mn mo mp mq mr ms ij bi translated">与用运算符“|”捕获多个异常不同，在这个例子中，我对异常使用了多个捕获。这可能会使我的代码更长，但是根据我得到的结果，正如你在例子中看到的，我可以确定我的代码发生了什么异常。例如:</p><figure class="kp kq kr ks gt kt"><div class="bz fp l di"><div class="nx ny l"/></div><figcaption class="la lb gj gh gi lc ld bd b be z dk translated">具有多个 Catch 块的多个异常示例</figcaption></figure><h1 id="a800" class="lf lg iq bd lh li lj lk ll lm ln lo lp kf lq kg lr ki ls kj lt kl lu km lv lw bi translated">#3 记录并丢弃</h1><p id="6288" class="pw-post-body-paragraph lx ly iq lz b ma mb ka mc md me kd mf mg mh mi mj mk ml mm mn mo mp mq mr ms ij bi translated">log 和 throw 有区别也有相似之处。参见下面的例子，可以更清楚地了解什么时候应该使用 log，什么时候应该使用 throw。</p><h2 id="fc91" class="nm lg iq bd lh nn no dn ll np nq dp lp mg nr ns lr mk nt nu lt mo nv nw lv iw bi translated">用日志处理异常</h2><p id="c922" class="pw-post-body-paragraph lx ly iq lz b ma mb ka mc md me kd mf mg mh mi mj mk ml mm mn mo mp mq mr ms ij bi translated">下面的例子向我们展示了如何在 catch 块中使用 log。每当我们使用日志时，我们都可以在应用程序日志中获得关于异常的信息，而不会终止。当您希望您的应用程序运行时，尽管您知道会发生异常，但只想忽略这些异常以继续其他正常进程时，通常会使用这种方法。例如:</p><figure class="kp kq kr ks gt kt"><div class="bz fp l di"><div class="nx ny l"/></div><figcaption class="la lb gj gh gi lc ld bd b be z dk translated">用日志示例处理异常</figcaption></figure><p id="5766" class="pw-post-body-paragraph lx ly iq lz b ma mv ka mc md mw kd mf mg nj mi mj mk nk mm mn mo nl mq mr ms ij bi translated">您可以看到下面的输出:</p><figure class="kp kq kr ks gt kt"><div class="bz fp l di"><div class="nx ny l"/></div><figcaption class="la lb gj gh gi lc ld bd b be z dk translated">使用日志示例处理异常的输出</figcaption></figure><h1 id="98d9" class="lf lg iq bd lh li lj lk ll lm ln lo lp kf lq kg lr ki ls kj lt kl lu km lv lw bi translated">用 Throw 处理异常</h1><p id="4a4e" class="pw-post-body-paragraph lx ly iq lz b ma mb ka mc md me kd mf mg mh mi mj mk ml mm mn mo mp mq mr ms ij bi translated">有时我们不接受一些可以通过 try catch 块传递的异常。这是因为例外可能是关键问题或重要的业务不应该错过。在这种情况下，我们应该像下面的例子那样使用 throw 让应用程序被异常终止。例如:</p><figure class="kp kq kr ks gt kt"><div class="bz fp l di"><div class="nx ny l"/></div><figcaption class="la lb gj gh gi lc ld bd b be z dk translated">使用抛出示例处理异常</figcaption></figure><p id="8e26" class="pw-post-body-paragraph lx ly iq lz b ma mv ka mc md mw kd mf mg nj mi mj mk nk mm mn mo nl mq mr ms ij bi translated">代码的输出:</p><figure class="kp kq kr ks gt kt"><div class="bz fp l di"><div class="nx ny l"/></div><figcaption class="la lb gj gh gi lc ld bd b be z dk translated">带有抛出示例的句柄异常的输出</figcaption></figure><h1 id="42db" class="lf lg iq bd lh li lj lk ll lm ln lo lp kf lq kg lr ki ls kj lt kl lu km lv lw bi translated">#4 Finally 块和返回语句</h1><p id="b8cb" class="pw-post-body-paragraph lx ly iq lz b ma mb ka mc md me kd mf mg mh mi mj mk ml mm mn mo mp mq mr ms ij bi translated">try 块中带有 return 语句的 finally 块有一个特例。下面的例子将向你展示这是如何工作的。</p><p id="d230" class="pw-post-body-paragraph lx ly iq lz b ma mv ka mc md mw kd mf mg nj mi mj mk nk mm mn mo nl mq mr ms ij bi translated">正如您在代码中看到的，您可以获取并运行代码，然后您可能会看到代码的结果是“1”而不是“0”。这是因为当我们在 try 块中有一个 return 语句并且在 try 之后有一个 finally 块时，在执行 return 语句之前，代码应该首先执行 finally 中的所有代码。所以这里的第一个 return 语句是“return 1”，而不是“0”。</p><p id="d9b3" class="pw-post-body-paragraph lx ly iq lz b ma mv ka mc md mw kd mf mg nj mi mj mk nk mm mn mo nl mq mr ms ij bi translated">例如:</p><figure class="kp kq kr ks gt kt"><div class="bz fp l di"><div class="nx ny l"/></div><figcaption class="la lb gj gh gi lc ld bd b be z dk translated">Finally 块和返回语句示例</figcaption></figure><h1 id="b55d" class="lf lg iq bd lh li lj lk ll lm ln lo lp kf lq kg lr ki ls kj lt kl lu km lv lw bi translated">使用 Finally Block 的 5 个技巧</h1><p id="b766" class="pw-post-body-paragraph lx ly iq lz b ma mb ka mc md me kd mf mg mh mi mj mk ml mm mn mo mp mq mr ms ij bi translated">我想展示的与异常主题相关的最后一件事是使用 finally 块时的提示，您可以在实际情况中应用或避免这些提示。</p><h2 id="99f9" class="nm lg iq bd lh nn no dn ll np nq dp lp mg nr ns lr mk nt nu lt mo nv nw lv iw bi translated">不使用 Catch 块传递异常</h2><p id="f1f5" class="pw-post-body-paragraph lx ly iq lz b ma mb ka mc md me kd mf mg mh mi mj mk ml mm mn mo mp mq mr ms ij bi translated">在某些情况下，我们可能会放置一个 try 块来传递一些我们认为代码可能会发生的未知异常。有时我们只需要传递，而 catch 块中没有我们想要实现的东西。在这种情况下，我们可以跳过 catch 块，只使用 try with finally，如下例所示:</p><figure class="kp kq kr ks gt kt"><div class="bz fp l di"><div class="nx ny l"/></div><figcaption class="la lb gj gh gi lc ld bd b be z dk translated">传递异常，但不传递 Catch 块示例</figcaption></figure><h2 id="61b8" class="nm lg iq bd lh nn no dn ll np nq dp lp mg nr ns lr mk nt nu lt mo nv nw lv iw bi translated">不应在 Finally 块中引发任何异常</h2><p id="e743" class="pw-post-body-paragraph lx ly iq lz b ma mb ka mc md me kd mf mg mh mi mj mk ml mm mn mo mp mq mr ms ij bi translated">实际上，我们可以在任何我们想编码的地方编码。这个例子并不完全要求你继续下去。这只是一个小小的警告，当您使用 throw with finally 块时，可能会使您无法获得原来的异常。相反，您只会看到您在 finally 块中抛出的异常。例如:</p><figure class="kp kq kr ks gt kt"><div class="bz fp l di"><div class="nx ny l"/></div><figcaption class="la lb gj gh gi lc ld bd b be z dk translated">不建议对 Finally 块示例使用 Throw</figcaption></figure><h1 id="dffb" class="lf lg iq bd lh li lj lk ll lm ln lo lp kf lq kg lr ki ls kj lt kl lu km lv lw bi translated">结论</h1><p id="9389" class="pw-post-body-paragraph lx ly iq lz b ma mb ka mc md me kd mf mg mh mi mj mk ml mm mn mo mp mq mr ms ij bi translated">在本文中，我们探讨了 Java 异常处理的一些最佳实践。尽管决定使用哪种方法有许多因素，但本文中分享的最佳实践应该为您提供一个很好的切入点，以便您进行更多的探索或将其应用到您未来的 Java 项目中。我希望这些最佳实践对您有所帮助。</p></div><div class="ab cl nz oa hu ob" role="separator"><span class="oc bw bk od oe of"/><span class="oc bw bk od oe of"/><span class="oc bw bk od oe"/></div><div class="ij ik il im in"><p id="b8bf" class="pw-post-body-paragraph lx ly iq lz b ma mv ka mc md mw kd mf mg nj mi mj mk nk mm mn mo nl mq mr ms ij bi translated"><em class="og">喜欢这篇文章？给我弄个</em> <a class="ae le" href="https://ko-fi.com/techisbeautiful" rel="noopener ugc nofollow" target="_blank"> <em class="og"> Ko-fi </em> </a> <em class="og">。</em></p><p id="e2dc" class="pw-post-body-paragraph lx ly iq lz b ma mv ka mc md mw kd mf mg nj mi mj mk nk mm mn mo nl mq mr ms ij bi translated"><em class="og">爱我的文字？加入我的</em> <a class="ae le" href="https://medium.com/subscribe/@techisbeautiful" rel="noopener"> <em class="og">邮箱列表</em> </a> <em class="og">。</em></p><p id="5577" class="pw-post-body-paragraph lx ly iq lz b ma mv ka mc md mw kd mf mg nj mi mj mk nk mm mn mo nl mq mr ms ij bi translated"><em class="og">爱读书？加入</em><a class="ae le" href="https://medium.com/@techisbeautiful/membership" rel="noopener"><em class="og">Medium</em></a><em class="og">(如果你用上面的链接，也是支持我，因为我有 Medium 的小提成)。</em></p></div><div class="ab cl nz oa hu ob" role="separator"><span class="oc bw bk od oe of"/><span class="oc bw bk od oe of"/><span class="oc bw bk od oe"/></div><div class="ij ik il im in"><p id="5bf7" class="pw-post-body-paragraph lx ly iq lz b ma mv ka mc md mw kd mf mg nj mi mj mk nk mm mn mo nl mq mr ms ij bi translated">如果你喜欢这个故事，你也会喜欢:</p><ul class=""><li id="b822" class="mt mu iq lz b ma mv md mw mg mx mk my mo mz ms na nb nc nd bi translated"><a class="ae le" rel="noopener ugc nofollow" target="_blank" href="/java-18-top-4-features-you-must-know-1f36ee23e2ab">你必须知道的 Java 18: 4 特性</a></li><li id="f904" class="mt mu iq lz b ma ne md nf mg ng mk nh mo ni ms na nb nc nd bi translated"><a class="ae le" href="https://medium.com/@techisbeautiful/java-17-top-5-features-you-must-know-bbed2afaea3d" rel="noopener">你必须知道的 Java 17: 5 特性</a></li><li id="4a79" class="mt mu iq lz b ma ne md nf mg ng mk nh mo ni ms na nb nc nd bi translated"><a class="ae le" href="https://medium.com/@techisbeautiful/new-features-you-must-know-in-java-11-and-examples-3fda2ad26def" rel="noopener">你必须知道的 Java 11: 8 特性</a></li><li id="8c7c" class="mt mu iq lz b ma ne md nf mg ng mk nh mo ni ms na nb nc nd bi translated"><a class="ae le" rel="noopener ugc nofollow" target="_blank" href="/java-8-seven-features-you-must-know-and-examples-1c3964ae7fe8">你必须知道的 Java 8 : 7 特性</a></li></ul></div></div>    
</body>
</html>